# HG changeset patch
# User Carlo de Falco <cdf@users.sourceforge.net>
# Date 1370432987 -7200
#      Wed Jun 05 13:49:47 2013 +0200
# Node ID c8bbab6b9e7a4ee111c9f46a6bc736c27f58fb4a
# Parent  a71a784c7611e13076685fb97d2250a95c04bc3e
Avoid converting sparse matrix to full format in svds.

* script/sparse/svds.m: fix a bug where the matrix passed as
  sparse in input would be unnecessarily converted to full storage
  for computing the maximum magnitude of its nonzeros.

diff --git a/scripts/sparse/svds.m b/scripts/sparse/svds.m
--- a/scripts/sparse/svds.m
+++ b/scripts/sparse/svds.m
@@ -132,17 +132,21 @@ function [u, s, v, flag] = svds (A, k, s
     if (sigma < 0)
       error ("svds: SIGMA must be a positive real value");
     endif
   else
     error ("svds: SIGMA must be a positive real value or the string 'L'");
   endif
 
   [m, n] = size (A);
-  max_a = max (abs (A(:)));
+  max_a = max (abs (nonzeros (A)));
+  if (isempty (max_a))
+    max_a = 0;
+  endif
+
   if (max_a == 0)
     s = zeros (k, 1);  # special case of zero matrix
   else
     if (nargin < 2)
       k = min ([6, m, n]);
     else
       k = min ([k, m, n]);
     endif
